api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,307 @@
+
+    // The server must send back this message in response to a message with
+    // the "request_trailers" field set.
+    TrailersResponse request_trailers = 5;
+
+    // The server must send back this message in response to a message with
+    // the "response_trailers" field set.
+    TrailersResponse response_trailers = 6;
+
+    // If specified, attempt to create a locally generated response, send it
+    // downstream, and stop processing additional filters and ignore any
+    // additional messages received from the remote server for this request or
+    // response. If a response has already started -- for example, if this
+    // message is sent response to a "response_body" message -- then
+    // this will either ship the reply directly to the downstream codec,
+    // or reset the stream.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Optional metadata that will be emitted as dynamic metadata to be consumed by the next
+  // filter. This metadata will be placed in the namespace "envoy.filters.http.ext_proc".
+  google.protobuf.Struct dynamic_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 2;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 3;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to an HttpHeaders message.
+message HeadersResponse {
+  CommonResponse response = 1;
+}
+
+// This message must be sent in response to an HttpTrailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the trailers
+  HeaderMutation header_mutation = 1;
+}
+
+// This message must be sent in response to an HttpBody message.
+message BodyResponse {
+  CommonResponse response = 1;
```

Comment:
Why do `HeadersReponse` and `BodyResponse` devolve to the same `CommonResponse`? It seems that they would have different structure, e.g. you can't modify headers in body response after headers have been sent.

Reply:
Well sometimes you can and sometimes you can't. You can't modify headers in the body response after they have been sent but I think that you can if you buffered the body (at least that's something I want to support, so that an ext_proc server can modify everything together.) You could also in theory decide in a body callback to ignore the rest of the body from the upstream and downstream and ignore the headers and replace the rest of the HTTP message with your own thing -- that's what the CONTINUE_AND_DONT_END_STREAM status is supposed to do.

Reply:
It seems that we're linearizing Envoy's internal filter callbacks here, since HTTP filters essentially treat headers and body as independent event streams. Is this by design?

Reply:
The protocol tries to keep the request and response streams separate -- even though there are separate HttpHeaders and HttpBody messages, they're broken into separate fields like request_headers and response_headers for that reason. So the design is not to linearize that.

However, I am assuming that the request or response headers will come before the body, and I'm planning to use "StopAllIterationAndWatermark" so that the ext_proc server has a chance to respond to the request and response headers callbacks before being asked to present the message. If that's not valid then I'm not sure what to do, because again I'm sure that we'll want the ability to modify the headers and body together sometimes.

Reply:
Make sense. Looking at the request flow, I'm trying to understand how this works though. Let's say you have sent to the server two messages, one with headers, and the next with body. How many response messages are expected? One for both the headers and body request message? One just for the body request message? Would an empty headers response message be sent when you want to do processing in the body?

Do we have a strict request-response protocol that doesn't allow pipelining? Would it be more efficient for Envoy to be able to fire the headers and body requests without feedback?

Reply:
That's what's documented right now -- Envoy sends a "request_headers" message, pauses the filter chain, and waits for a "request_headers" response. That seemed like the most straightforward thing for an ext_proc server writer to support -- then the server can respond with a new processing mode if it wants the body messages, or not. Otherwise it's already too late to respond I think.

If we want to support other modes, we can add more options to the ProcessingMode message or fine-tune our interpretation there.

Reply:
It's probably fine to start with this, but I'd be curious what it would take to reduce the round trips in a later API. When folks start optimizing latency of ext_proc based requests, this is probably going to be low hanging fruit.